services/api/service.go

```diff
@@ -493,6 +527,85 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
+	t := time.Now()
+	requestErr, validationErr = api.blockSimRateLimiter.Send(ctx, opts.req, opts.isHighPrio, opts.fastTrack)
+	log := opts.log.WithFields(logrus.Fields{
+		"duration":   time.Since(t).Seconds(),
+		"numWaiting": api.blockSimRateLimiter.CurrentCounter(),
+	})
+	if validationErr != nil {
+		ignoreError := validationErr.Error() == ErrBlockAlreadyKnown || validationErr.Error() == ErrBlockRequiresReorg || strings.Contains(validationErr.Error(), ErrMissingTrieNode)
+		if !ignoreError {
+			log.WithError(validationErr).Warn("block validation failed")
+			return nil, validationErr
+		}
+		log.WithError(validationErr).Warn("ignorable error during block simulation")
+		return nil, nil
+	}
+	if requestErr != nil {
+		log.WithError(requestErr).Warn("block validation failed: request error")
+		return requestErr, nil
+	}
+	log.Info("block validation successful")
+	return nil, nil
+}
+
+func (api *RelayAPI) demoteBuilder(pubkey string, req *common.BuilderSubmitBlockRequest, simError error) {
+	builderEntry, ok := api.blockBuildersCache[pubkey]
+	if !ok {
+		api.log.Warnf("builder %v not in the builder cache", pubkey)
+		builderEntry = &blockBuilderCacheEntry{} //nolint:exhaustruct
+	}
+	newStatus := common.BuilderStatus{
+		IsHighPrio:    builderEntry.status.IsHighPrio,
+		IsBlacklisted: builderEntry.status.IsBlacklisted,
+		IsOptimistic:  false,
+	}
+	api.log.Infof("demoted builder, new status: %v", newStatus)
+	if err := api.db.SetBlockBuilderStatus(pubkey, newStatus, true); err != nil {
+		api.log.Error(fmt.Errorf("error setting builder: %v status: %w", pubkey, err))
+	}
+	// Write to demotions table.
+	api.log.WithFields(logrus.Fields{"builder_pubkey": pubkey}).Info("demoting builder")
+	if err := api.db.InsertBuilderDemotion(req, simError); err != nil {
+		api.log.WithError(err).WithFields(logrus.Fields{
+			"errorWritingDemotionToDB": true,
+			"bidTrace":                 req.Message,
+			"simError":                 simError,
+		}).Error("failed to save demotion to database")
+	}
+}
+
+// processOptimisticBlock is called on a new goroutine when a optimistic block
+// needs to be simulated.
+func (api *RelayAPI) processOptimisticBlock(opts blockSimOptions) {
+	api.optimisticBlocksInFlight.Add(1)
+	defer func() { api.optimisticBlocksInFlight.Sub(1) }()
+	api.optimisticBlocksWG.Add(1)
+	defer api.optimisticBlocksWG.Done()
+
+	ctx := context.Background()
+	builderPubkey := opts.req.BuilderPubkey().String()
+	opts.log.WithFields(logrus.Fields{
+		"builderPubkey": builderPubkey,
+		// NOTE: this value is just an estimate because many goroutines could be
+		// updating api.optimisticBlocksInFlight concurrently. Since we just use
+		// it for logging, it is not atomic to avoid the performance impact.
+		"optBlocksInFlight": api.optimisticBlocksInFlight,
+	}).Infof("simulating optimistic block with hash: %v", opts.req.BuilderSubmitBlockRequest.BlockHash())
+	reqErr, simErr := api.simulateBlock(ctx, opts)
+	if reqErr != nil || simErr != nil {
```

Comment:
also mark as non-optimistic on a request error?

Reply:
yes i think so! we don't see request errors often enough to be an issue, and this is an extra failsafe so we can check the demotion and make sure the block is there before repromoting. again this is a relatively rare event.